In [64]:
import pandas as pd

## Checking What does the data look like?

In [65]:
df = pd.read_csv("../data/train.csv")
df.head()


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [66]:
df.shape

(1460, 81)

In [67]:
df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [68]:
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning             str
LotFrontage      float64
LotArea            int64
                  ...   
MoSold             int64
YrSold             int64
SaleType             str
SaleCondition        str
SalePrice          int64
Length: 81, dtype: object

## Checking for missing values

In [69]:
missing_values_percentage = df.isnull().sum()/len(df) * 100
missing_values_percentage = missing_values_percentage[missing_values_percentage > 0].sort_values(ascending=False)
print(missing_values_percentage)

PoolQC          99.520548
MiscFeature     96.301370
Alley           93.767123
Fence           80.753425
MasVnrType      59.726027
FireplaceQu     47.260274
LotFrontage     17.739726
GarageType       5.547945
GarageYrBlt      5.547945
GarageFinish     5.547945
GarageQual       5.547945
GarageCond       5.547945
BsmtExposure     2.602740
BsmtFinType2     2.602740
BsmtQual         2.534247
BsmtCond         2.534247
BsmtFinType1     2.534247
MasVnrArea       0.547945
Electrical       0.068493
dtype: float64


In [70]:
df.drop(missing_values_percentage[missing_values_percentage > 50].index, axis=1, inplace=True)

In [71]:
print(missing_values_percentage)

PoolQC          99.520548
MiscFeature     96.301370
Alley           93.767123
Fence           80.753425
MasVnrType      59.726027
FireplaceQu     47.260274
LotFrontage     17.739726
GarageType       5.547945
GarageYrBlt      5.547945
GarageFinish     5.547945
GarageQual       5.547945
GarageCond       5.547945
BsmtExposure     2.602740
BsmtFinType2     2.602740
BsmtQual         2.534247
BsmtCond         2.534247
BsmtFinType1     2.534247
MasVnrArea       0.547945
Electrical       0.068493
dtype: float64


In [72]:
print(missing_values_percentage)

PoolQC          99.520548
MiscFeature     96.301370
Alley           93.767123
Fence           80.753425
MasVnrType      59.726027
FireplaceQu     47.260274
LotFrontage     17.739726
GarageType       5.547945
GarageYrBlt      5.547945
GarageFinish     5.547945
GarageQual       5.547945
GarageCond       5.547945
BsmtExposure     2.602740
BsmtFinType2     2.602740
BsmtQual         2.534247
BsmtCond         2.534247
BsmtFinType1     2.534247
MasVnrArea       0.547945
Electrical       0.068493
dtype: float64


## Handling Missing Values 

### Here I'm Answering 3 questions to guide on how to handle the missing values:
1. **Does the data dictionary explain the NA? if YES --> I follow what it says**
2. **Does the missing Value mean "Does not exist"?if yes --> fill with "None" for Categorical, 0 numeric**
3. **Is it genuinelly uncollected data with no pattern?if yes --> This is MCAR, I can use median for numeric, mode(most frequent) for categorical**


| Column | % Missing | Type | Decision | Reason |
|--------|-----------|------|----------|--------|
| FireplaceQu | 47.26%| Categorical | Fill "None" | Data dict says NA = No Fireplace |
| LotFrontage | 17.74%| Numeric     | Fill median |  MCAR — genuinely uncollected    |
| GarageType  | 5.55% | Categorical | Fill "None" | Data dict says NA = No Garage    |
| GarageYrBlt | 5.55% | Numeric     | Fill 0      | Data dict says NA = No Garage    |
| GarageFinish| 5.55% | Categorical | Fill "None" | Data dict says NA = No Garage    |
| GarageQual  | 5.55% | Categorical | Fill "None" | Data dict says NA = No Garage    |
| GarageCond  | 5.55% | Categorical | Fill "None" | Data dict says NA = No Garage    |
| BsmtExposure| 2.60% | Categorical | Fill "None" | Data dict says NA = No Basement  |
| BsmtFinType2| 2.60% | Categorical | Fill "None" | Data dict says NA = No Basement  |
| BsmtQual    | 2.53% | Categorical | Fill "None" | Data dict says NA = No Basement  |
| BsmtCond    | 2.53% | Categorical | Fill "None" | Data dict says NA = No Basement  |
| BsmtFinType1| 2.53% | Categorical | Fill "None" | Data dict says NA = No Basement  |
| MasVnrArea  | 0.55% | Numeric     | Fill 0      | MCAR — genuinely uncollected     |
| Electrical  | 0.07% | Categorical | Fill mode   | MCAR — genuinely uncollected     |


In [73]:
all_values_counts = df["Electrical"].value_counts()
print(all_values_counts)

Electrical
SBrkr    1334
FuseA      94
FuseF      27
FuseP       3
Mix         1
Name: count, dtype: int64


In [74]:
df["LotFrontage"].describe()

count    1201.000000
mean       70.049958
std        24.284752
min        21.000000
25%        59.000000
50%        69.000000
75%        80.000000
max       313.000000
Name: LotFrontage, dtype: float64

## Code to handle all 3 case of the 14 columns that have null values

In [75]:
df.isnull().sum()[df.isnull().sum() > 0]

LotFrontage     259
MasVnrArea        8
BsmtQual         37
BsmtCond         37
BsmtExposure     38
BsmtFinType1     37
BsmtFinType2     38
Electrical        1
FireplaceQu     690
GarageType       81
GarageYrBlt      81
GarageFinish     81
GarageQual       81
GarageCond       81
dtype: int64

### Handling the Null values using mode, median, NAN where data dict say NAN and zero for numerical that are missing completely at random(MCAR)

In [78]:
none_columns =  ["BsmtQual", "BsmtCond", "BsmtExposure", "BsmtFinType1", "BsmtFinType2", "FireplaceQu", "GarageType", "GarageFinish", "GarageQual", "GarageCond"]
zero_columns = ["MasVnrArea","GarageYrBlt"]

for col in none_columns:
    df[col] = df[col].fillna("None")


for col in zero_columns:
    #print(col)
    df[col] = df[col].fillna(0)

median_LotFrontage = df["LotFrontage"].median()
df["LotFrontage"] = df["LotFrontage"].fillna(median_LotFrontage)

mode_Electrical = df["Electrical"].mode()[0]
df["Electrical"] = df["Electrical"].fillna(mode_Electrical)




In [79]:
df.isnull().sum()[df.isnull().sum() > 0] #df.isnull().sum()[df.isnull().sum() > 0]

Series([], dtype: int64)